## The R scripts and conda environments used in the analysis for the paper of "Transformation of brain myeloid cell populations by SIV in rhesus macaques revealed by multiomics"
### The R packages in this study include: 
1. [ArchR(version: 1.0.2)](http://www.archrproject.com/)
2. [Seurat(version: 4.4.0)](https://github.com/satijalab/seurat)
3. [Monocle3(version: 1.3.7)](https://github.com/cole-trapnell-lab/monocle3)
4. [cellchat(version: 2.1.0)](https://github.com/jinworks/CellChat)
5. [ClusterProfiler(version: 4.0.2)](https://bioconductor.org/packages/clusterProfiler/)
6. [SCENIC(version: 1.1.2-01)](https://htmlpreview.github.io/?https://github.com/aertslab/SCENIC/blob/master/inst/doc/SCENIC_Running.html)

### This study has been published on [Commun Biol 8, 100 (2025). Xu, X., Niu, M., Lamberty, B.G. et al.](https://doi.org/10.1038/s42003-024-07443-4)Transformation of brain myeloid cell populations by SIV in rhesus macaques revealed by multiomics.
### Abstract
The primary immune constituents in the brain, microglia and macrophages, are the target for HIV in people and simian immunodeficiency virus (SIV) in nonhuman primates. This infection can lead to neurological dysfunction, known as HIV-associated neurocognitive disorder (HAND). Given the gaps in our knowledge on how these cells respond in vivo to CNS infection, we perform single-cell multiomic sequencing, including gene expression and ATAC-seq, on myeloid cells from the brains of rhesus macaques with SIV-induced encephalitis (SIVE) as well as uninfected controls. We find that SIVE significantly changes the myeloid cell populations. In SIVE, microglia-like cells express high levels of chemoattractants capable of recruiting highly activated CAM-like cells to the site of infection/inflammation. A unique population of microglia-like cells is found in which the chromatin accessibility of genes diverges from their RNA expression. Additionally, we observe a dramatic shift of upstream gene regulators and their targets in brain myeloid cells during SIVE. This study further uncovers the transcriptome, gene regulatory events, and potential roles of different brain myeloid phenotypes in SIVE. This might deepen the understanding of SIVE/HIVE and enlighten the therapeutic development.
![](https://github.com/Howard-Fox-Lab/SIVE_Multiomic_analysis/blob/main/abstract_graph.png)
